From 340832e88bd1e23aa01c4559d9680dba606924c9 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:13 -0400
Subject: [fs] nfsd: ensure that st_access_bmap and st_deny_bmap are initialized to 0

Message-id: <1410898849-22126-70-git-send-email-bfields@redhat.com>
Patchwork-id: 95710
O-Subject: [Patch RHEL7 069/165] nfsd: ensure that st_access_bmap and st_deny_bmap are initialized to 0
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@primarydata.com>

Open stateids must be initialized with the st_access_bmap and
st_deny_bmap set to 0, so that nfs4_get_vfs_file can properly record
their state in old_access_bmap and old_deny_bmap.

This bug was introduced in commit baeb4ff0e502 (nfsd: make deny mode
enforcement more efficient and close races in it) and was causing the
refcounts to end up incorrect when nfs4_get_vfs_file returned an error
after bumping the refcounts. This made it impossible to unmount the
underlying filesystem after running pynfs tests that involve deny modes.

Signed-off-by: Jeff Layton <jlayton@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 2f6ce8e73caa443201e3d826639b9242cf6ea568
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index a3a828d..66a3b84 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -2970,8 +2970,6 @@ static void init_open_stateid(struct nfs4_ol_stateid *stp, struct nfs4_file *fp,
  stp->st_file = fp;
  stp->st_access_bmap = 0;
  stp->st_deny_bmap = 0;
- set_access(open->op_share_access, stp);
- set_deny(open->op_share_deny, stp);
  stp->st_openstp = NULL;
  spin_lock(&fp->fi_lock);
  list_add(&stp->st_perfile, &fp->fi_stateids);
-- 
1.7.1